spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -488,123 +563,155 @@ NOTE: It is up to individual implementations how they will provide access-contro
+    upgrade = provableStore.get(channelUpgradePath(portIdentifier, channelIdentifier))
 
-    // counterparty must be in TRY state
-    if counterpartyChannel.State != TRYUPGRADE {
-        restoreChannel(portIdentifier, channelIdentifier)
-        return
+    // in the crossing hellos case, the versions returned by both on TRY must be the same
+    if currentChannel.state == TRYUPGRADE {
+        if upgrade.fields.version != counterpartyUpgrade.fields.version {
+            restoreChannel(portIdentifier, channelIdentifier)
+        }
     }
 
-    // both channel ends must be mutually compatible.
-    // this means that the ordering must be the same and 
-    // any future introduced fields that must be compatible
-    // should also be checked
-    if counterpartyChannel.ordering != proposedUpgradeChannel.ordering {
-        restoreChannel(portIdentifier, channelIdentifier)
-        return
-    }
+    // prove counterparty and move our own state to ACKUPGRADE and start flushing
+    // upgrade is blocked on this channelEnd from progressing until flush completes on both ends
+    startFlushUpgradeHandshake(portIdentifier, channelIdentifier, upgrade.fields, counterpartyChannel, counterpartyUpgrade, ACKUPGRADE, proofChannel, proofUpgrade, proofHeight)
 
     // call modules onChanUpgradeAck callback
+    // module can error on counterparty version
+    // ACK should not change state to the new parameters yet
+    // as that will happen on the onChanUpgradeOpen callback
     module = lookupModule(portIdentifier)
     err = module.onChanUpgradeAck(
         portIdentifier,
         channelIdentifier,
-        counterpartyChannel.channelIdentifier,
-        counterpartyChannel.version
+        counterpartyUpgrade.version
     )
     // restore channel if callback returned error
     if err != nil {
         restoreChannel(portIdentifier, channelIdentifier)
         return
     }
 
-    // upgrade is complete
-    // set channel to OPEN and remove unnecessary state
-    currentChannel.state = OPEN
-    provableStore.set(channelPath(portIdentifier, channelIdentifier), currentChannel)
-    provableStore.delete(channelUpgradeTimeoutPath(portIdentifier, channelIdentifier))
-    privateStore.delete(channelRestorePath(portIdentifier, channelIdentifier))
+    // if no error, agree on final version
+    upgrade.version = counterpartyUpgrade.version
+    provableStore.set(channelUpgradePath(portIdentifier, channelIdentifier), upgrade)
+
+    // refresh channel
+    currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+
+    // if both sides have already flushed then open the upgrade handshake immediately
+    if  currentChannel.state == FLUSHCOMPLETE && counterpartyFlushStatus == FLUSHCOMPLETE {
+        openChannelHandshake(portIdentifier, channelIdentifier)
+        module.onChanUpgradeOpen(portIdentifier, channelIdentifier)
+    }
 }
 ```
 
 ```typescript
-function chanUpgradeConfirm(
+function chanUpgradeOpen(
```

Comment:
Is there good reason to stray from the `Confirm` naming convention? I find it a bit more natural, as open is kind of close to `Init` in some ways

Reply:
I guess since the channel state is changing to `OPEN` if the in flight packets have been flushed, it might be confusing to have the function called `Confirm` when the end state will actually `OPEN`?